File: app/src/main/java/no/joharei/flixr/modules/MainModule.java
package no.joharei.flixr.modules;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import no.joharei.flixr.api.AndroidFlickrApi;
import no.joharei.flixr.api.FlickrApiContainer;
import no.joharei.flixr.api.FlickrInterceptor;
import no.joharei.flixr.api.LocalCredentialStore;
import no.joharei.flixr.api.models.AuthToken;
import no.joharei.flixr.utils.ApiKeys;
import no.joharei.flixr.utils.Constants;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import se.akerfeldt.okhttp.signpost.OkHttpOAuthConsumer;
import se.akerfeldt.okhttp.signpost.SigningInterceptor;

@Module
public class MainModule {

    @Provides
    FlickrApiContainer provideApiContainer(OkHttpClient okHttpClient) {
        return new AndroidFlickrApi(okHttpClient, Constants.BASE_URL);
    }

    @Provides
    @Singleton
    OkHttpClient provideOkHttpClient(LocalCredentialStore localCredentialStore) {
        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder();

        clientBuilder.addInterceptor(new FlickrInterceptor());

        OkHttpOAuthConsumer consumer = new OkHttpOAuthConsumer(ApiKeys.CONSUMER_KEY, ApiKeys.CONSUMER_SECRET);
        AuthToken authToken = localCredentialStore.getToken();
        // TODO: refresh somehow?
        consumer.setTokenWithSecret(authToken.getAuthToken(), authToken.getAuthTokenSecret());
        clientBuilder.addInterceptor(new SigningInterceptor(consumer));

        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor();
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        clientBuilder.addInterceptor(httpLoggingInterceptor);

        return clientBuilder.build();
    }
}

File: app/src/main/java/no/joharei/flixr/modules/MainModule.kt
package no.joharei.flixr.modules

import com.google.gson.FieldNamingPolicy
import com.google.gson.GsonBuilder
import dagger.Module
import dagger.Provides
import no.joharei.flixr.api.AndroidFlickrApi
import no.joharei.flixr.api.FlickrApiContainer
import no.joharei.flixr.api.FlickrInterceptor
import no.joharei.flixr.api.LocalCredentialStore
import no.joharei.flixr.common.ApiKeys
import no.joharei.flixr.common.Constants
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.converter.gson.GsonConverterFactory
import se.akerfeldt.okhttp.signpost.OkHttpOAuthConsumer
import se.akerfeldt.okhttp.signpost.SigningInterceptor
import javax.inject.Singleton


@Module
class MainModule {

    @Provides
    fun provideApiContainer(okHttpClient: OkHttpClient, gsonConverterFactory: GsonConverterFactory): FlickrApiContainer {
        return AndroidFlickrApi(okHttpClient, Constants.BASE_URL, gsonConverterFactory)
    }

    @Provides
    @Singleton
    fun provideOkHttpClient(consumer: OkHttpOAuthConsumer): OkHttpClient {
        val clientBuilder = OkHttpClient.Builder()
        clientBuilder.addInterceptor(FlickrInterceptor())
        clientBuilder.addInterceptor(SigningInterceptor(consumer))

        val httpLoggingInterceptor = HttpLoggingInterceptor()
        httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
        clientBuilder.addInterceptor(httpLoggingInterceptor)

        return clientBuilder.build()
    }

    @Provides
    @Singleton
    fun provideGsonConverterFactory(): GsonConverterFactory {
        val gson = GsonBuilder()
                .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                .create()
        return GsonConverterFactory.create(gson)
    }

    @Provides
    @Singleton
    fun provideOkHttpOAuthConsumer(localCredentialStore: LocalCredentialStore): OkHttpOAuthConsumer {
        val consumer = OkHttpOAuthConsumer(ApiKeys.CONSUMER_KEY, ApiKeys.CONSUMER_SECRET)
        val authToken = localCredentialStore.token
        consumer.setTokenWithSecret(authToken.authToken, authToken.authTokenSecret)
        return consumer
    }
}
